File: nmLabReport/lab17.c
#include <stdio.h>
#include <math.h>

#define MAX 100
#define EPSILON 0.001

void gaussSeidel(float a[MAX][MAX], float b[MAX], float x[MAX], int n) {
    float x_new[MAX];
    int i, j, k;
    float sum;

    for (i = 0; i < n; i++) {
        x[i] = 0;
    }

    for (k = 0; k < MAX; k++) {
        for (i = 0; i < n; i++) {
            sum = b[i];
            for (j = 0; j < n; j++) {
                if (i != j) {
                    sum -= a[i][j] * x[j];
                }
            }
            x_new[i] = sum / a[i][i];
        }

        int converged = 1;
        for (i = 0; i < n; i++) {
            if (fabs(x_new[i] - x[i]) > EPSILON) {
                converged = 0;
                break;
            }
        }

        if (converged) {
            break;
        }

        for (i = 0; i < n; i++) {
            x[i] = x_new[i];
        }
    }
}

int main() {
    int n, i, j;
    float a[MAX][MAX], b[MAX], x[MAX];

    printf("Enter the number of equations: ");
    scanf("%d", &n);

    printf("Enter the coefficients of the matrix A:\n");
    for (i = 0; i < n; i++) {
        for (j = 0; j < n; j++) {
            scanf("%f", &a[i][j]);
        }
    }

    printf("Enter the constants vector B:\n");
    for (i = 0; i < n; i++) {
        scanf("%f", &b[i]);
    }

    gaussSeidel(a, b, x, n);

    printf("The solution is:\n");
    for (i = 0; i < n; i++) {
        printf("x[%d] = %f\n", i, x[i]);
    }

    return 0;
}

File: nmLabReport/lab17.py
import numpy as np

def gauss_seidel(a, b, tol=1e-6, max_iter=100):
    n = len(a)
    x = np.zeros(n)
    for _ in range(max_iter):
        x_new = np.copy(x)
        for i in range(n):
            s = sum(a[i][j] * x_new[j] for j in range(n) if j != i)
            x_new[i] = (b[i] - s) / a[i][i]
        if np.allclose(x, x_new, atol=tol):
            break
        x = x_new
    return x

def main():
    n = int(input("Enter the number of equations: "))
    a = np.zeros((n, n))
    b = np.zeros(n)

    print("Enter the coefficients of the matrix A:")
    for i in range(n):
        a[i] = list(map(float, input().split()))

    print("Enter the constants vector B:")
    b = list(map(float, input().split()))

    x = gauss_seidel(a, b)

    print("The solution is:")
    for i in range(n):
        print(f"x[{i}] = {x[i]:.6f}")

if __name__ == "__main__":
    main()